[FEATURE] Ajout du néerlandais dans les langues disponibles côté API (PIX-10683)

reibecca opened the pull request:
## ❄️ Problème
Afin de préparer l'arrivée de Pix App en néerlandais, nous avons besoin de gérer cette nouvelle langue dans l'API.

## ⛄️ Proposition

- Extraction de l'objet qui gère les langues du fichier `api/src/shared/domain/constants.js`
- Création d'un nouveau service `language-service.js` où l'on va centraliser toute la gestion des langues 
- Ajout de la langue `DUTCH: 'nl'`
- Suite à la [suppression de la contrainte sur les langues en BDD](https://github.com/1024pix/pix/pull/7828), ajout de la gestion d'erreur des langues disponibles dans ce nouveau service
- Ajout de langue par défaut `Français` directement dans le model `User.js` grâce au `language-service.js`
- Utilisation d'une variable `AVAILABLE_LANGUAGES` dans la validation Joi des routes, pour éviter de devoir modifier plusieurs fichiers lors du prochain ajout d'une nouvelle langue

## ☕️ Remarques
RAS.

## 🍫 Pour tester

### 🟡 Pix Admin 

#### ✅ Cas de succès
- Sur Postman, se connecter avec le compte `superadmin@example.net`
- Faire un appel `PATCH` sur la route `/api/admin/users/10002` avec comme payload : 

```
{
    "data": {
        "id": "10002",
        "attributes": {
            "first-name": "NextSuper",
            "last-name": "NextAdmin",
            "lang": "nl"
        }
    }
}
```
- Vous devez avoir une réponse avec un `statut 200` et un objet avec la langue de l'utilisateur définit à : `"lang": "nl"`

#### ❌ Cas d'erreurs
- Sur Postman, se connecter avec le compte `superadmin@example.net`
- Faire un appel `PATCH` sur la route `/api/admin/users/10002` avec comme payload : 

```
{
    "data": {
        "id": "10002",
        "attributes": {
            "first-name": "NextSuper",
            "last-name": "NextAdmin",
            "lang": "it"
        }
    }
}
```
- Vous devez avoir une réponse avec un `statut 400` et le message `"\"data.attributes.lang\" must be one of [en, fr, nl]"`

#### 🆒 Tests de non regression 
- Dans le payload, remplacer la langue par ` "lang": "en"` ou ` "lang": "fr"` et vérifier que vous avez bien une réponse avec un `statut 200` et un objet avec la langue de l'utilisateur définit à : ` "lang": "en"` ou ` "lang": "fr"`.

<HR>

### 🟣 Pix App

#### ✅ Cas de succès
- Sur Postman, se connecter avec un compte utilisateur, ex: `camille-onette@example.net`
- Faire un appel `PATCH` sur la route `/api/users/101670/lang/nl` : 
- Vous devez avoir une réponse avec un `statut 200` et un objet avec la langue de l'utilisateur définit à : `"lang": "nl"`


#### ❌ Cas d'erreurs
- Sur Postman, se connecter avec un compte utilisateur, ex: `camille-onette@example.net`
- Faire un appel `PATCH` sur la route `/api/users/101670/lang/jp` : 
- Vous devez avoir une réponse avec un `statut 400` et le message `"\"lang\" must be one of [en, fr, nl]"`

#### 🆒 Tests de non regression 
- Dans votre appel, remplacer la langue par ` /api/users/101670/lang/en` ou ` /api/users/101670/lang/fr` et vérifier que vous avez bien une réponse avec un `statut 200` et un objet avec la langue de l'utilisateur définit à : ` "lang": "en"` ou ` "lang": "fr"`.

reibecca committed:
refactor(api): add context in tests

Co-authored-by: Ismael Gorissen <ismael.gorissen@gmail.com>
Co-authored-by: Eric Lim <eric.lim@zenika.com>
Co-authored-by: Emmanuelle Bonnemay <emmanuelle.bonnemay@pix.fr>

pix-bot-github commented:
Une fois les applications déployées, elles seront accessibles via les liens suivants :
  * [App (.fr)](https://app-pr7838.review.pix.fr)
  * [App (.org)](https://app-pr7838.review.pix.org)
  * [Orga (.fr)](https://orga-pr7838.review.pix.fr)
  * [Orga (.org)](https://orga-pr7838.review.pix.org)
  * [Certif (.fr)](https://certif-pr7838.review.pix.fr)
  * [Certif (.org)](https://certif-pr7838.review.pix.org)
  * [1D](https://1d-pr7838.review.pix.fr)
  * [Admin](https://admin-pr7838.review.pix.fr)
  * [API](https://api-pr7838.review.pix.fr/api/)
  * [Audit Logger](https://pix-audit-logger-review-pr7838.osc-fr1.scalingo.io/api/)

Les variables d'environnement seront accessibles via les liens suivants :
  * [scalingo front](https://dashboard.scalingo.com/apps/osc-fr1/pix-front-review-pr7838/environment)
  * [scalingo api](https://dashboard.scalingo.com/apps/osc-fr1/pix-api-review-pr7838/environment)
  * [scalingo audit-logger](https://dashboard.scalingo.com/apps/osc-fr1/pix-audit-logger-review-pr7838/environment)

reibecca committed:
feat(api): create language service
- to handle new language : Dutch
- to handle errors in the code instead of the BDD

Co-authored-by: Ismael Gorissen <ismael.gorissen@gmail.com>
Co-authored-by: Eric Lim <eric.lim@zenika.com>
Co-authored-by: Emmanuelle Bonnemay <emmanuelle.bonnemay@pix.fr>

reibecca committed:
feat(api): use language service in User model

Co-authored-by: Ismael Gorissen <ismael.gorissen@gmail.com>
Co-authored-by: Eric Lim <eric.lim@zenika.com>
Co-authored-by: Emmanuelle Bonnemay <emmanuelle.bonnemay@pix.fr>

er-lim commented on api/src/shared/domain/errors.js:
```diff
@@ -111,6 +111,14 @@ class MissingAssessmentId extends DomainError {
 }
 
+class LanguageNotSupportedError extends DomainError {
+  constructor(languageCode) {
+    super(`Given language is not supported : "${languageCode}"`);
+    this.code = 'LANGUAGE_NOT_SUPPORTED';
+    this.meta = { languageCode };
+  }
+}
+
```
Je pense qu'il faut aussi ajouter le mapping en erreur HTTP dans le fichier error-manager.js comme ce qui est fait pour le [LocaleNotSupportedError](https://github.com/1024pix/pix/blob/dev/api/src/shared/application/error-manager.js#L104-L106)

igorissen replied:
Même si, potentiellement cette erreur ne peut pas arriver dû au faite que l'on a une validation Joi à l'entrée, c'est toujours bien de l'avoir.

er-lim commented on api/tests/shared/unit/domain/services/language-service_test.js:
```diff
@@ -0,0 +1,36 @@
+describe('Unit | Shared | Domain | Services | Language Service', function () {
+  describe('#assertLanguageAvailability', function () {
+    context('when language code is not provided', function () {
+      it('returns nothing (undefined)', function () {
```
```suggestion
      it('returns nothing ', function () {
```
Pour avoir la même description que le test en bas 😄

er-lim commented:
Test en RA sur Firefox ✅ 
Je suis passé par la console navigateur au lieu de Postman. Ca a bien marché aussi 😄

EmmanuelleBonnemay reviewed (approved):
Testé sur Postman avec succès
